api/controllers/shabads.js

```diff
@@ -509,7 +451,81 @@ const getShabad = (req, res, ShabadIDQ, sinceDate = null, forceMulti = false) =>
       .catch(err => reject(err));
   });
 
+const getAngs = async (req, res, { pageNo, sinceDate, sourceId }) => {
+  let PageNo = pageNo;
+  let SourceID = sourceId;
+  if (!lib.isRangeOfNumbers(PageNo)) {
+    PageNo = '1';
+  }
+
+  // Check if SourceID is supported or default to 'G'
+  if (!sources[SourceID]) {
+    SourceID = 'G';
+  }
+  // If SGGS, check if within 1430
+  if (SourceID === 'G' && PageNo > 1430) {
+    PageNo = '1430';
+  }
+
+  const PageNoQuery = lib.searchOperators.angToQuery(PageNo);
+
+  const parameters = [...PageNoQuery.parameters, SourceID];
+
+  let sinceQuery = '';
+  if (sinceDate) {
+    sinceQuery = 'AND GREATEST(s.Updated, v.Updated) > ?';
+    parameters.push(sinceDate);
+  }
+
+  let conn;
+  let results = null;
+  try {
+    conn = await req.app.locals.pool.getConnection();
+    const q = `SELECT ${allColumns} ${allFrom}
+      WHERE
+        ${PageNoQuery.q}
+        AND v.SourceID = ?
+        ${sinceQuery}
+        ${allColumnsWhere}
+      ORDER BY PageNo,v.LineNo ASC, ShabadID ASC, v.ID ASC`;
+
+    const rows = await conn.query(q, parameters);
+    if (rows.length > 0 && PageNoQuery.totalPages === 1) {
+      // single ang
+      const output = await getAngSingle(req, res, rows);
+      results = output;
+    } else if (rows.length > 0) {
+      // multiple ang
+      const output = {
+        pageNos: [],
+        pages: [],
+      };
+      let curPage = -1;
+      let counter = 0;
+      rows.forEach(row => {
+        if (row.PageNo !== curPage) {
+          curPage = row.PageNo;
+          output.pageNos.push(curPage);
+          counter = output.pages.push([]) - 1;
+        }
+        output.pages[counter].push(row);
+      });
+
+      const outputPagePromises = output.pages.map(row => getAngSingle(req, res, row));
+      output.pages = await Promise.all(outputPagePromises);
```

Comment:
Maybe we could use Promise.allSettled since it doesnt short curcuit if one fails